internal/fs/fs.go

```diff
@@ -2058,6 +2061,60 @@ func (fs *fileSystem) checkDirNotEmpty(dir inode.BucketOwnedDirInode, name strin
 	return nil
 }
 
+// Rename an old folder to a new folder in hierarchical bucket. If the new folder already
+// exists and is non-empty, return ENOTEMPTY.
+//
+// LOCKS_EXCLUDED(fs.mu)
+// LOCKS_EXCLUDED(oldParent)
+// LOCKS_EXCLUDED(newParent)
+func (fs *fileSystem) renameFolder(ctx context.Context, oldParent inode.DirInode, oldName string, newParent inode.DirInode, newName string) (err error) {
+	var pendingInodes []inode.DirInode
+	defer fs.releaseInodes(&pendingInodes)
+
+	oldDir, err := fs.getBucketDirInode(ctx, oldParent, oldName)
+	if err != nil {
+		return err
+	}
+	pendingInodes = append(pendingInodes, oldDir)
+
+	if err = fs.ensureNoOpenFilesInDirectory(oldDir, oldName); err != nil {
+		return err
+	}
+
+	newDir, err := fs.getBucketDirInode(ctx, newParent, newName)
+	if err == nil {
+		// If the directory exists, then check if it is empty or not.
+		if err = fs.checkDirNotEmpty(newDir, newName); err != nil {
+			return err
+		}
+		pendingInodes = append(pendingInodes, newDir)
+	}
+
+	oldDirName := inode.NewDirName(oldParent.Name(), oldName)
+	newDirName := inode.NewDirName(newParent.Name(), newName)
+	if oldParent == newParent {
+		// If both parents are the same, lock once.
+		oldParent.Lock()
+		defer oldParent.Unlock()
+	} else {
+		oldParent.Lock()
+		defer oldParent.Unlock()
+
+		newParent.Lock()
+		defer newParent.Unlock()
+	}
+
+	// Rename old directory to the new directory, keeping both parent directory locked.
+	_, err = oldParent.RenameFolder(ctx, oldDirName.GcsObjectName(), newDirName.GcsObjectName())
+	if err != nil {
+		return err
+	}
+
+	fs.releaseInodes(&pendingInodes)
+	fs.checkInvariantsForImplicitDirs()
```

Comment:
I am guessing that this should also be a deferred statement.

Reply:
Actually I thought we can remove this `fs.releaseInodes(&pendingInodes)` as we already doing it in defer above, but wanted to make sure with @vadlakondaswetha about importance of putting it twice.